hasura/graphql-engine
# unlock scheduled events on graceful shutdown

- When shutting down gracefully, unlock the locked scheduled events




### Description



The graphql-engine locks the scheduled events when it attempts to process it. Earlier, if a shutdown process was initiated the locked scheduled events remain locked forever. This PR, will unlock the locked events when graceful shutdown is initiated. 

If the graphql-engine has crashed, then the scheduled events will remain locked. To solve that, on startup, we unlock all the locked scheduled events, so that they can be processed.

### Changelog


### Affected components



### Related Issues




### Solution and Design


Maintain a set of the scheduled event ID's in the cache that have been fetched from the database and are in-line to be processed and when the graceful-shutdown is initiated, then use the ID's to update the status of the locked scheduled events.

### Steps to test and verify


Create lots of scheduled events with the same `schedule_at` and when the graphql-engine is processing them, initiate a graceful-shutdown. Then, check the status of the scheduled events, none of the scheduled events should have status as 'locked'.


### Limitations, known bugs & workarounds



### Server checklist


#### Catalog upgrade

Does this PR change Hasura Catalog version?

#### Metadata


Does this PR add a new Metadata feature?
  - Does `run_sql` auto manages the new metadata through schema diffing?
  - Does `run_sql` auto manages the definitions of metadata on renaming?
  - Does `export_metadata`/`replace_metadata` supports the new metadata added?


#### GraphQL
   
   

#### Breaking changes


  1. Metadata API

     Existing `query` types:
     

  2. GraphQL API

     Schema Generation:
     
     
     

     Schema Resolve:-
     
     

  3. Logging
